examples/python/cancellation/search.py

```diff
@@ -0,0 +1,158 @@
+        if min_distance is None or distance < min_distance:
+            min_distance = distance
+    return min_distance
+
+
+def _get_hash(secret):
+    hasher = hashlib.sha1()
+    hasher.update(secret)
+    return base64.b64encode(hasher.digest()).decode('ascii')
+
+
+class ResourceLimitExceededError(Exception):
+    """Signifies the request has exceeded configured limits."""
+
+
+def _bytestrings_of_length(length):
+    """Generates a stream containing all bytestrings of a given length.
+
+    Args:
+      length: A positive integer length.
+
+    Yields:
+      All bytestrings of length `length`.
+    """
+    digits = [0] * length
+    while True:
+        yield b''.join(struct.pack('B', i) for i in digits)
+        digits[-1] += 1
+        i = length - 1
+        while digits[i] == _BYTE_MAX + 1:
+            digits[i] = 0
+            i -= 1
+            if i == -1:
+                # Terminate the generator since we've run out of strings of
+                # `length` bytes.
+                raise StopIteration()  # pylint: disable=stop-iteration-return
+            else:
+                digits[i] += 1
+
+
+def _all_bytestrings():
+    """Generates a stream containing all possible bytestrings.
+
+    This generator does not terminate.
+
+    Yields:
+      All bytestrings in ascending order of length.
+    """
+    length = 1
+    while True:
+        for bytestring in _bytestrings_of_length(length):
+            yield bytestring
+        length += 1
+
+
+def search(target,
+           ideal_distance,
+           stop_event,
+           maximum_hashes,
+           interesting_hamming_distance=None):
+    """Find candidate strings.
+
+    Search through the space of all bytestrings, in order of increasing length,
+    indefinitely, until a hash with a Hamming distance of `maximum_distance` or
+    less has been found.
+
+    Args:
+      target: The search string.
+      ideal_distance: The desired Hamming distance.
+      stop_event: An event indicating whether the RPC should terminate.
+      maximum_hashes: The maximum number of hashes to check before stopping.
+      interesting_hamming_distance: If specified, strings with a Hamming
+        distance from the target below this value will be yielded.
+
+    Yields:
+      Instances  of HashNameResponse. The final entry in the stream will be of
+        `maximum_distance` Hamming distance or less from the target string,
+        while all others will be of less than `interesting_hamming_distance`.
+
+    Raises:
+      ResourceLimitExceededError: If the computation exceeds `maximum_hashes`
+        iterations.
+    """
+    hashes_computed = 0
+    for secret in _all_bytestrings():
+        if stop_event.is_set():
+            raise StopIteration()  # pylint: disable=stop-iteration-return
+        candidate_hash = _get_hash(secret)
+        distance = _get_substring_hamming_distance(candidate_hash, target)
```

Comment:
I would recommend to perform `lower` here instead of calling `lower` each time in `_get_hamming_distance`.